src/Features/Core/Portable/Debugging/DebugInformationReaderProvider.cs

```diff
@@ -0,0 +1,132 @@
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.DiaSymReader;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Debugging
+{
+    /// <summary>
+    /// An abstraction of a symbol reader that provides a reader of Edit and Continue debug information.
+    /// Owns the underlying PDB reader.
+    /// </summary>
+    internal abstract class DebugInformationReaderProvider : IDisposable
+    {
+        private sealed class DummySymReaderMetadataProvider : ISymReaderMetadataProvider
+        {
+            public static readonly DummySymReaderMetadataProvider Instance = new DummySymReaderMetadataProvider();
+
+            public unsafe bool TryGetStandaloneSignature(int standaloneSignatureToken, out byte* signature, out int length)
+                => throw ExceptionUtilities.Unreachable;
+
+            public bool TryGetTypeDefinitionInfo(int typeDefinitionToken, out string namespaceName, out string typeName, out TypeAttributes attributes)
+                => throw ExceptionUtilities.Unreachable;
+
+            public bool TryGetTypeReferenceInfo(int typeReferenceToken, out string namespaceName, out string typeName)
+                => throw ExceptionUtilities.Unreachable;
+        }
+
+        private sealed class Portable : DebugInformationReaderProvider
+        {
+            private readonly MetadataReaderProvider _pdbReaderProvider;
+
+            public Portable(MetadataReaderProvider pdbReaderProvider)
+                => _pdbReaderProvider = pdbReaderProvider;
+
+            public override EditAndContinueMethodDebugInfoReader CreateEditAndContinueMethodDebugInfoReader()
+                => EditAndContinueMethodDebugInfoReader.Create(_pdbReaderProvider.GetMetadataReader());
+
+            public override void Dispose()
+                => _pdbReaderProvider.Dispose();
+        }
+
+        private sealed class Native : DebugInformationReaderProvider
+        {
+            private readonly Stream _stream;
+            private readonly int _version;
+            private ISymUnmanagedReader5 _symReader;
+
+            public Native(Stream stream, ISymUnmanagedReader5 symReader, int version)
+            {
+                _stream = stream;
+                _symReader = symReader;
+                _version = version;
+            }
+
+            public override EditAndContinueMethodDebugInfoReader CreateEditAndContinueMethodDebugInfoReader()
+                => EditAndContinueMethodDebugInfoReader.Create(_symReader, _version);
+
+            public override void Dispose()
+            {
+                _stream.Dispose();
+
+                var symReader = Interlocked.Exchange(ref _symReader, null);
+                if (symReader != null && Marshal.IsComObject(symReader))
+                {
+                    Marshal.ReleaseComObject(symReader);
```

Comment:
By rule, all calls to this I must say I am terrified of.

Reply:
(Yes, I can imagine why this is safe. The rule stands. :smile:)

Reply:
That's understandable. Unfortunately this is how symreader needs to be disposed, otherwise it keeps the file open :(